src/Orleans.Serialization/Buffers/Writer.cs

```diff
@@ -349,8 +349,23 @@ private void WriteMultiSegment(scoped ReadOnlySpan<byte> source)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void WriteByte(byte value)
         {
-            EnsureContiguous(1);
-            _currentSpan[_bufferPos++] = value;
+            if ((uint)_bufferPos < (uint)_currentSpan.Length)
+            {
+                Unsafe.AddByteOffset(ref MemoryMarshal.GetReference(_currentSpan), (uint)_bufferPos) = value;
```

Comment:
This is not necessary, best codegen is when written as
```c#
public void WriteByte(byte value)
{
    int bufferPos = _bufferPos;
    if ((uint)bufferPos < (uint)_currentSpan.Length)
    {
        _currentSpan[bufferPos] = value;
        _bufferPos = bufferPos + 1;
    }
    else
    {
        WriteByteSlow(value);
    }
}
```
* there's no bound check emitted
* `_bufferPos` is read and written only once (memory accesses), while with `_bufferPos++` it's 2 reads + 1 write (for the `inc`)

Reply:
Your input is appreciated, @gfoidl

Reply:
There's still going to be a bound check unless `_currentSpan` is also first copied to a local variable, but that introduces additional overhead.

Reply:
Just curious: on what target platform did you see this? Is this after inlining?
Asking because I checked the codegen and didn't see the bound check -- so the JIT should be taught to eliminate that one.

Reply:
I tested on x64 and after inlining.

Reply:
Well this seems to be https://github.com/dotnet/runtime/issues/72004 (actually opened by myself, and I forgot about this 😉).

Reply:
Perhaps we should link to that issue and this comment from code so that it can be re-assessed later